### PR TITLE
handle partially quantized models

### PR DESCRIPTION
**Pull request by davidkoski**
- fix for #53 #71 #69 #74
- in order to test the models
	- I added a default prompt of an appropriate form
	- while working on the model configuration also added additional stop tokens (#74)
- fixed the repetitionPenalty code (#71)

Note: this isn't directly usable as it requires https://github.com/ml-explore/mlx-swift/pull/73 (I was using the local checkout for development).  This will need an update to the mlx version after 73 is merged.

### CODE REVIEW

File: Applications/LLMEval/ContentView.swift [hunk 1]
```diff
@@ -125,6 +125,8 @@ struct ContentView: View {
 
         }
         .task {
+            self.prompt = llm.modelConfiguration.defaultPrompt
```
Comment (davidkoski) [inline Applications/LLMEval/ContentView.swift:128]: Use the new default prompt (in case the model changes)

File: Libraries/LLM/Evaluate.swift [hunk 2]
```diff
@@ -12,7 +12,7 @@ private func topPSampling(logits: MLXArray, topP: Float, temp: Float) -> MLXArra
         logits = logits.asType(.float32)
     }
 
-    let probs = softMax(logits / temp, axis: -1)
+    let probs = softmax(logits / temp, axis: -1)
```
Comment (davidkoski) [inline Libraries/LLM/Evaluate.swift:15]: deprecation warning fix

File: Libraries/LLM/Evaluate.swift [hunk 3]
```diff
@@ -120,9 +120,8 @@ public struct TokenIterator: Sequence, IteratorProtocol {
         y = sample(logits: logits, temp: parameters.temperature, topP: parameters.topP)
         // append the current token to the context and check repetitionPenalty context see if need to remove the first token
         if parameters.repetitionContextSize > 1 {
-            repetitionContext = concatenated([repetitionContext, y], axis: 0)
             if repetitionContext.shape[0] > parameters.repetitionContextSize {
-                repetitionContext = repetitionContext[1...]
+                repetitionContext = repetitionContext[(-parameters.repetitionContextSize)...]
```
Comment (davidkoski) [inline Libraries/LLM/Evaluate.swift:124]: Fix for #71 -- this uses a direct port of the python code.  Previously it looks like there were workarounds for the lack of full numpy array indexing, see `take()` above.

File: Libraries/LLM/Evaluate.swift [hunk 4]
```diff
@@ -174,14 +173,32 @@ public enum GenerateDisposition {
...
+                tokenizer.encode(text: $0)
+            }
+            .filter {
+                // discard anything that is not a single token.  sometimes
+                // the tokenizer will insert a <s> token, so accept that too
+                $0.count == 1 || ($0.count == 2 && $0[0] == 1)
+            }
+            .map {
+                $0.last!
+            })
```
Comment (davidkoski) [inline Libraries/LLM/Evaluate.swift:200]: Fix for #74

File: Libraries/LLM/Load.swift [hunk 5]
```diff
@@ -54,9 +54,15 @@ public func load(
...
 
+        // per-model cleanup
+        weights = model.sanitize(weights: weights)
+
         // quantize if needed
         if let quantization = baseConfig.quantization {
-            quantizeIfNeeded(model: model, weights: weights, quantization: quantization)
+            quantize(model: model, groupSize: quantization.groupSize, bits: quantization.bits) {
+                path, module in
+                weights["\(path).scales"] != nil
```
Comment (davidkoski) [inline Libraries/LLM/Load.swift:64]: This is the fix for loading partially quantized models -- if they have "scales" then the layer is quantized.  We don't need to test the type of layer as the `quantize()` method will only convert layers that _can_ be converted.

File: Libraries/LLM/Load.swift [hunk 6]
```diff
@@ -76,38 +82,3 @@ public func load(
             hub: hub, configuration: newConfiguration, progressHandler: progressHandler)
     }
 }
-
-// MARK: - Quantization
-
-private func quantizeIfNeeded(
-    model: LLMModel, weights: [String: MLXArray], quantization: BaseConfiguration.Quantization
-) {
```
Comment (davidkoski) [inline Libraries/LLM/Load.swift:84]: No longer needed.

File: Libraries/LLM/Load.swift (hunk 5 above)
Comment (davidkoski) [reply Libraries/LLM/Load.swift:64]: Note that this depends on https://github.com/ml-explore/mlx-swift/pull/73

File: Libraries/LLM/Models.swift [hunk 7]
```diff
@@ -98,11 +112,16 @@ public struct ModelConfiguration {
...
+
+        // https://www.promptingguide.ai/models/mistral-7b
+        defaultPrompt: "describe the swift language"
+    )
 
     public static let codeLlama13b4bit = ModelConfiguration(
         id: "mlx-community/CodeLlama-13b-Instruct-hf-4bit-MLX",
-        overrideTokenizer: "PreTrainedTokenizer"
+        overrideTokenizer: "PreTrainedTokenizer",
+        defaultPrompt: "func sortArray(_ array: [Int]) -> String { <FILL_ME> }"
```
Comment (davidkoski) [inline Libraries/LLM/Models.swift:124]: I verified each of these models and added a defaultPrompt so I don't have to hunt around each time

File: Libraries/LLM/Models.swift [hunk 8]
```diff
@@ -111,40 +130,53 @@ extension ModelConfiguration {
         "<PRE> " + prompt.replacingOccurrences(of: "<FILL_ME>", with: "<SUF>") + " <MID>"
     }
 
-    public static let phi4bit = ModelConfiguration(id: "mlx-community/phi-2-hf-4bit-mlx") {
-        prompt in
-        "Instruct: \(prompt)\nOutput: "
```
Comment (davidkoski) [inline Libraries/LLM/Models.swift:116]: This wasn't giving good results (any more?) and isn't used on the python side.

File: Libraries/LLM/Qwen2.swift [hunk 9]
```diff
@@ -249,6 +271,8 @@ public struct Qwen2Configuration: Codable {
                 Bool.self, forKey: Qwen2Configuration.CodingKeys.ropeTraditional) ?? false
         self.ropeScaling = try container.decodeIfPresent(
             [String: StringOrNumber].self, forKey: Qwen2Configuration.CodingKeys.ropeScaling)
+        self.tieWordEmbeddings =
+            try container.decodeIfPresent(Bool.self, forKey: .tieWordEmbeddings) ?? false
```
Comment (davidkoski) [inline Libraries/LLM/Qwen2.swift:275]: The qwen2 model implementation has changed slightly

**Comment by davidkoski** [comment]
note: the CI tests are failing because this depends on the mlx-swift change that needs to merge first

**Comment by solume** [comment]
tested it with mlx-swift fork [](https://github.com/ml-explore/mlx-swift/pull/73), this breaks for me with Libraries/LLM/Load.swift:62:13 Cannot find 'quantize' in scope
and
Libraries/LLM/Qwen2.swift:198:37 Value of type 'Embedding' has no member 'asLinear'

**Comment by solume** [comment]
this still gives me an error even after linking the correct branch:
```
LLM >
Missing required module '_NumericsShims'
```

**Comment by solume** [comment]
was able to fix dependencies but getting error:
`libc++abi: terminating due to uncaught exception of type std::invalid_argument: [matmul] Last dimension of first input with shape (1,42,1280) must match second to last dimension of second input with shape (160,32000).`

File: Libraries/LLM/Load.swift (hunk 6 above)
Comment (awni) [reply Libraries/LLM/Load.swift:84]: Nice simplification!

File: Libraries/LLM/Qwen2.swift [hunk 10]
```diff
@@ -191,8 +193,26 @@ public class Qwen2Model: Module, LLMModel {
...
+        }
+        return (out, cache)
+    }
+
+    public func sanitize(weights: [String: MLXArray]) -> [String: MLXArray] {
+        var weights = weights
+
+        if configuration.tieWordEmbeddings && weights["lm_head.weight"] == nil {
+            weights["lm_head.weight"] = nil
+        }
```
Comment (awni) [inline Libraries/LLM/Qwen2.swift:210]: This looks kind of odd. If it's equal to `nil`, why set it to `nil`? Is that a Swift thing?

File: Libraries/LLM/Models.swift (hunk 7 above)
Comment (awni) [reply Libraries/LLM/Models.swift:124]: Beautiful, thanks!

File: Libraries/LLM/Evaluate.swift [hunk 11]
```diff
@@ -174,14 +173,32 @@ public enum GenerateDisposition {
 ///   - parameters: generation parameters
 ///   - model: model to evaluate
 ///   - tokenizer: tokenizer to convert tokens back into strings and recognizer special tokens
+///   - configuration: the model configuration
 ///   - didGenerate: visitor for the tokens as they are generated
 public func generate(
     promptTokens: [Int], parameters: GenerateParameters, model: LLMModel, tokenizer: Tokenizer,
+    configuration: ModelConfiguration,
```
Comment (awni) [inline Libraries/LLM/Evaluate.swift:180]: I'm not certain about the way this model configuration gets passed around and what the intention for it is. Somehow we don't have a need for it in Python, so I'm wondering why do you need it here? What should go in it vs in the tokenizer/model directly?

From what I can tell it's more like the default arguments for a given model (eos token / prompt). The prompt gets handled outside this function. So is here it's just for the additional eos tokens?

File: Libraries/LLM/Qwen2.swift (hunk 10 above)
Comment (davidkoski) [reply Libraries/LLM/Qwen2.swift:210]: No, you are right that is odd.

Comment (davidkoski) [reply Libraries/LLM/Qwen2.swift:210]: Here is the original python:

```python
        if self.args.tie_word_embeddings:
            weights.pop("lm_head.weight", None)
```

I think I can just remove the `&& ...`

File: Libraries/LLM/Evaluate.swift (hunk 11 above)
Comment (davidkoski) [reply Libraries/LLM/Evaluate.swift:180]: yeah, just for the additional eos tokens.  I can switch that to just pass in additional eos tokens (optional).  If we need more parameters in the future I can rethink that slightly.

**Review by awni — APPROVED** [review]
Thanks!! LGTM

**Pull request merged**